src/fswatch_win/fswatch_win_stubs.c

```diff
@@ -0,0 +1,317 @@
+  DWORD name_len;
+  WCHAR *file_name;
+
+  while (events != NULL) {
+    info = events->info;
+    v_directory = caml_copy_string(events->dirpath);
+
+    for (;;) {
+      name_len = info->FileNameLength / sizeof(WCHAR);
+      file_name = caml_stat_alloc(sizeof(WCHAR) * (name_len + 1));
+      memcpy(file_name, info->FileName, sizeof(WCHAR) * name_len);
+      file_name[name_len] = 0;
+      v_path = caml_copy_string_of_utf16(file_name);
+      caml_stat_free(file_name);
+
+      /* Allocate Event.t */
+      ev = caml_alloc_tuple(3);
+      Store_field(ev, 0, v_directory);
+      Store_field(ev, 1, v_path);
+      Store_field(ev, 2, Val_action(info->Action));
+
+      /* Allocate list cell */
+      cell = caml_alloc_tuple(2);
+      Store_field(cell, 0, ev);
+      Store_field(cell, 1, res);
+      res = cell;
+
+      if (info->NextEntryOffset == 0)
+        break;
+
+      info = (FILE_NOTIFY_INFORMATION *)((char*)info + info->NextEntryOffset);
+    }
+
+    next = events->next;
+    caml_stat_free(events->info);
+    caml_stat_free(events);
+    events = next;
+  }
+
+  CAMLreturn(res);
+}
+
+value fswatch_win_create(value v_unit) {
+  CAMLparam1(v_unit);
+  struct fsenv *fsenv = malloc(sizeof(struct fsenv));
+  HANDLE iocp, signal;
+  DWORD err;
+
+  if ((iocp = CreateIoCompletionPort(INVALID_HANDLE_VALUE, NULL, 0, 1)) == NULL) {
+    unix_error("CreateIoCompletionPort", GetLastError());
+  }
+
+  if ((signal = CreateEvent(NULL, FALSE, FALSE, NULL)) == NULL) {
+    err = GetLastError();
+    CloseHandle(iocp);
+    unix_error("CreateEvent", err);
+  }
+
+  fsenv->events = NULL;
+  fsenv->iocp = iocp;
+  fsenv->signal = signal;
+
+  CAMLreturn(caml_copy_nativeint((intnat)fsenv));
+}
+
+value fswatch_win_wait(value v_fsenv) {
+  CAMLparam1(v_fsenv);
+  struct fsenv *fsenv = (struct fsenv *)Nativeint_val(v_fsenv);
+  DWORD res;
+
+  caml_release_runtime_system();
+  res = WaitForSingleObject(fsenv->signal, INFINITE);
+  caml_acquire_runtime_system();
+
+  switch (res) {
+    case WAIT_OBJECT_0:
+      break;
+    case WAIT_FAILED:
+      unix_error("WaitForSingleObject", GetLastError());
+      break;
+    default:
+      CAMLassert(0);
+      break;
+  }
+
+  CAMLreturn(Val_unit);
+}
+
+value fswatch_win_get_events(value v_fsenv) {
+  struct fsenv *fsenv = (struct fsenv *)Nativeint_val(v_fsenv);
+  struct events *events = pop_events(fsenv);
+  ResetEvent(fsenv->signal);
+  return parse_events(events);
+}
+
+#define FILE_NOTIFY_CONDITIONS \
+  (FILE_NOTIFY_CHANGE_FILE_NAME | FILE_NOTIFY_CHANGE_DIR_NAME | FILE_NOTIFY_CHANGE_LAST_WRITE)
+
+static void read_changes(struct wenv *wenv) {
+  memset(wenv->overlapped, 0, sizeof(OVERLAPPED));
+  wenv->buffer = caml_stat_alloc(FILE_NOTIFY_BUFFER_LENGTH);
+  ReadDirectoryChangesW(wenv->dir_handle, wenv->buffer, FILE_NOTIFY_BUFFER_LENGTH, TRUE,
+                        FILE_NOTIFY_CONDITIONS, NULL, wenv->overlapped, NULL);
+}
+
+static DWORD __stdcall watch_thread(LPVOID param) {
```

Comment:
ocaml/ocaml uses the macro `WINAPI` instead, which expands to `__stdcall`. It is ignored in arches other than x86 ([doc](https://learn.microsoft.com/en-us/cpp/cpp/stdcall?view=msvc-170)).
 
```suggestion
static DWORD WINAPI watch_thread(LPVOID param) {
```

Reply:
Good idea, fixed.